### ISSUE TITLE
Add the ability to pass any objects as signal binds, not just built-in types

### ISSUE THREAD

**Issue by xfactor2000** [godotengine/godot#16706]
**Godot version:**
3.0.stable.official.mono

**OS/device including version:**
Manjaro Linux 64 bit (rolling release)

**Issue description:**
I am unable to pass custom classes as signal binds, getting exception. I understand this is expected behavior, but I would like to be able to pass any objects alongside my signals.

**Steps to reproduce:**

See the attached project. When clicked, the button generates an exception:

0:00:00:0782 - Attempted to convert an unmarshallable managed type to Variant. Name: 'CustomClass' Encoding: 18

Attempted to convert an unmarshallable managed type to Variant. Name: 'CustomClass' Encoding: 18
Time: 0:00:00:0782
C Error: Method/Function Failed, returning: Variant()
C Source: modules/mono/mono_gd/gd_mono_marshal.cpp:593
C Function: mono_object_to_variant

**Minimal reproduction project:**
[TestProjectBindCustomClass.zip](https://github.com/godotengine/godot/files/1725364/TestProjectBindCustomClass.zip)

Please note that the weird .Connect function syntax (object[] within object[] is due to this bug:

#16311

**Comment by Duehok** [comment]
This feature would be very very nice to have.

[The delegate/event/eventArgs pattern of C#](https://docs.microsoft.com/en-us/dotnet/csharp/event-pattern) can be used for an equivalent but is more complex, not integrated in the editor (and harder to integrate than the signals) and I have not yet tested how it interacts with free_queue.

What's worse is that the error is a runtime error instead of a compile-time error.

**Comment by xfactor2000** [comment]
Duehok, your delegate solution seems valid although I need to try it. However, I still believe objects in signal binds should be implemented at language-agnostic level.

**Comment by vnen** [comment]
The problem is that Godot is not aware of anything that doesn't inherit its Object class (in fact, signals are a property of the Godot objects). It seems inconvenient but it's intended to be used as a script language so all your game classes should be Godot objects (which internally can call anything offered by the language).

Maybe @neikeq can find some workaround, but IMO this doesn't need to work this way.

**Comment by xfactor2000** [comment]
I tried your approach by extending CustomClass from Node or Object classes, then I get another error:

> ERROR: emit_signal: Error calling method from signal 'pressed': 'Node2D(game.cs):: _on_Btn_pressed': Method not found.
   At: core/object.cpp:1202.

**Comment by vnen** [comment]
@xfactor2000 yeah, that one is probably a bug.

**Comment by xfactor2000** [comment]
@vnen So should I do something, like poking somebody?

**Comment by Fido789** [comment]
Godot's inability to pass custom objects to signals is really limitating. A workaround to this issue I have found is to use a custom wrapper class for signal parameters. Unfortunatelly generic wrapper doesn't work, so it must be something like this:

```
public class ParameterWrapper : Godot.Object
    {
        private readonly object value;

        public ParameterWrapper(object value)
        {
            this.value = value;
        }

        public T GetValue<T>()
        {
            return (T)this.value;
        }
    }
```

Could any of core developers confirm, that this is a valid solution and will not produce memory leaks or something?

**Comment by xfactor2000** [comment]
Nope, in fact it works just fine, it's just that your custom objects have
to inherit from Node. See the discussion above.

**Comment by Fido789** [comment]
That's great. So in that case couln't be such a wrapper implemented at engine's level? Just checking for the type and if is is not built-in type or Godot.Object, wrap/unwrap it?

**Comment by xfactor2000** [comment]
Just pass a custom class that inherits from Node inside binds array, and
you will receive it as function parameter.

**Comment by Fido789** [comment]
Sure, I can do that, but a lot of my objects doesn't derive from Node/Godot.Object. For example my Level class. It would be very convenient if I could pass it as a parameter to level buttons. I can pass it wrapped in a wrapper above, but if it can be easily wrapped at the engine level, why not to do it?

**Comment by xfactor2000** [comment]
Why not make your level class inherit from Node? That's what I did, never
bothered me.

**Comment by Fido789** [comment]
Well there are many reasons, like memory consumption, inability to derive from anything else, easy transfer to another engine (right now Unity -> Godot), unit testing and many other. If I would do that, the cure would be much worse than symptoms.

**Comment by xfactor2000** [comment]
You can make your base class derive from Node and then derive from it. As
for memory consumption - why is it an issue? Node is meant to be a basic
building block of Godot, so it makes sense to inherit from it often. If it
was not optimized for wide usage, Godot wouldn't work.

**Comment by Fido789** [comment]
It is not that memory consumption would be huge or something, it is just unnecessary. For now I will stick to the wrapper.

**Comment by eddex** [comment]
@Fido789 thanks for the solution with the wrapper! This just made my day 😄

**Comment by rsoulard** [comment]
I am using the wrapper currently, but the reason I cannot just inherit from Godot.Object or Node is because the classes used come from a shared library that is used by other applications that do not "know" about Godot.

**Comment by KuroiRoy** [comment]
So I'm using the [ExtraMath](https://github.com/aaronfranke/GodotExtraMath) library and I'm trying to use a Vector4 in a signal. The wrapper object shared above works fine for this, but it is a little weird that you can't see what type your passing around. Inheriting from Godot.Object makes sense some sense when passing a Level or a Player because there aren't that many. Making a value type inherit from Godot.Object just to pass it around is a weird though.

For now I'll just make a signal wrapper specifically for Vector4 so you can see what type it is at a glance. But it would be nice if this was there was a way to wrap it at the engine level

**Comment by maximiliancsuk** [comment]
The problem with the wrapper approach is that it leaks. I was going through the godot.log in my projecct today and found the following log lines on exit:
```
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
   at: cleanup (core/object.cpp:2070) - ObjectDB instances leaked at exit (run with --verbose for details).
Leaked instance: Object:2344
Leaked instance: Object:2110
Leaked instance: Object:2227
[...snip...]
```

After some poking, I found that the wrapper node is responsible for this. Whenever you create a wrapper, it needs to be disposed. But I don't know how you could properly do that without a lot of boilerplate code. 

The wrapper approach is IMHO not good.

**Comment by neikeq** [comment]
The wrapper should inherit `Reference`/`RefCounted` for automatic collection.

**Comment by maximiliancsuk** [comment]
> The wrapper should inherit `Reference`/`RefCounted` for automatic collection.

You're right! That solves it, thanks for the tip!
